hnn_core/parallel_backends.py

```diff
@@ -345,105 +496,14 @@ def simulate(self, net):
-
-        # create the selector instance and register all input events
-        # with self.read_stdout which will only echo to stdout
-        self.sel = selectors.DefaultSelector()
-        self.sel.register(pipe_stdout_r, selectors.EVENT_READ,
-                          self._read_stdout)
-        self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
-                          self._read_stdout)
-
-        data_len = 0
-        completed = False
-        timeout = 0
-        # loop while the process is running
-        while True:
-            if not proc.poll() is None:
-                if completed is True:
-                    break
-                elif timeout > 4:
-                    # This is indicative of a failure. For debugging purposes.
-                    warn("Timed out (5s) waiting for end of data after child "
-                         "process stopped")
-                    break
-                else:
-                    timeout += 1
-                    sleep(1)
-
-            # wait for an event on the selector, timeout after 1s
-            events = self.sel.select(timeout=1)
-            for key, mask in events:
-                callback = key.data
-                completion_signal = callback(key.fileobj, mask)
-                if completion_signal is not None:
-                    if completion_signal == "end_of_sim":
-                        # finishied receiving printable output
-                        # everything else received is data
-                        self.sel.unregister(pipe_stderr_r)
-                        self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
-                                          self._read_stderr)
-                    elif isinstance(completion_signal, int):
-                        data_len = completion_signal
-                        self.sel.unregister(pipe_stdout_r)
-                        completed = True
-
-                        # there could still be data in stderr, so we return
-                        # to waiting until the process ends
-                    else:
-                        raise ValueError("Unrecognized signal received from "
-                                         "MPI child")
-
-        # cleanup the selector
-        self.sel.unregister(pipe_stderr_r)
-        self.sel.close()
-
-        # done with stdout and stderr
-        os.close(pipe_stdout_r)
-        os.close(pipe_stdout_w)
-        os.close(pipe_stderr_r)
-        os.close(pipe_stderr_w)
-
-        # if simulation failed, raise exception
-        if proc.returncode != 0:
-            raise RuntimeError("MPI simulation failed")
-
-        sim_data = self._process_child_data(self.proc_data_bytes, data_len)
+        proc, proc_data_bytes, data_len = run_subprocess(
+            command=self.mpi_cmd, pickled_obj=pickled_net, timeout=4,
+            env=env, cwd=os.getcwd(), universal_newlines=True)
+
+        sim_data = _process_child_data(proc_data_bytes, data_len)
```

Comment:
one of the main objectives of this redesign was to have as little code in `MPIBackend.simulate()` as in `JoblibBackend.simulate()`